File: src/test/java/org/vaadin/addons/ai/formfiller/utils/GoogleVisionLineSegmentationParser.java
package org.vaadin.addons.ai.formfiller.utils;

import com.google.cloud.vision.v1.AnnotateImageResponse;
import com.google.cloud.vision.v1.BoundingPoly;
import com.google.cloud.vision.v1.EntityAnnotation;
import com.google.cloud.vision.v1.Vertex;
import com.google.gson.Gson;

import java.awt.Polygon;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

// https://github.com/sshniro/line-segmentation-algorithm-to-gcp-vision/blob/master/kotlin/src/main/kotlin/com/google/vision/linesegmentation/GoogleVisionLineSegmentationParser.kt
// https://stackoverflow.com/questions/42391009/text-extraction-line-by-line

public class GoogleVisionLineSegmentationParser {

    @SuppressWarnings("unchecked")
    public List<String> initLineSegmentation(AnnotateImageResponse data) {
        int yMax = getYMax(data.getTextAnnotations(0));

        AnnotateImageResponse newData = invertAxis(data, yMax);
        List<String> lines = new ArrayList<>(Arrays.asList(newData.getTextAnnotationsList().get(0).getDescription().split("\n")));
        Collections.reverse(lines);

        List<EntityAnnotation> rawText = new ArrayList<>(newData.getTextAnnotationsList());
        Collections.reverse(rawText);
        rawText.remove(rawText.size() - 1);

        List<EntityAnnotation> mergedArray = getMergedLines(lines, rawText);
        List<Pair<EntityAnnotation, EntityMetadata>> entityToMetadata = getBoundingPolygon(mergedArray);

        combineBoundingPolygon(entityToMetadata);
        return constructLineWithBoundingPolygon(entityToMetadata);
    }

    protected List<String> constructLineWithBoundingPolygon(List<Pair<EntityAnnotation, EntityMetadata>> entityToMetadata) {
        ArrayList<String> finalArray = new ArrayList<>();
        for (int index = 0; index < entityToMetadata.size(); index++) {
            Pair<EntityAnnotation, EntityMetadata> it = entityToMetadata.get(index);
            if (!it.getSecond().matched) {
                if (it.getSecond().match.isEmpty()) {
                    finalArray.add(it.getFirst().getDescription());
                } else {
                    finalArray.add(arrangeWordsInOrder(entityToMetadata, index));
                }
            }
        }
        return finalArray;
    }

    private List<EntityAnnotation> getMergedLines(List<String> lines, List<EntityAnnotation> rawText) {
        ArrayList<EntityAnnotation> mergedArray = new ArrayList<>();
        while (lines.size() != 1) {
            String l = lines.remove(lines.size() - 1);
            String l1 = l;
            boolean status = true;
            EntityAnnotation mergedElement = null;

            while (true) {
                if (rawText.isEmpty()) {
                    break;
                }
                EntityAnnotation wElement = rawText.remove(rawText.size() - 1);
                String w = wElement.getDescription();

                int index = l.indexOf(w);

                l = l.substring(index + w.length());

                if (status) {
                    status = false;
                    mergedElement = wElement;
                }
                if (l.equals("")) {
                    EntityAnnotation newElement = EntityAnnotation.newBuilder().mergeFrom(mergedElement)
                            .setDescription(l1)
                            .setBoundingPoly(BoundingPoly.newBuilder().mergeFrom(mergedElement.getBoundingPoly()).clearVertices()
                                    .addVertices(0, mergedElement.getBoundingPoly().getVerticesList().get(0))
                                    .addVertices(1, wElement.getBoundingPoly().getVerticesList().get(1))
                                    .addVertices(2, wElement.getBoundingPoly().getVerticesList().get(2))
                                    .addVertices(3, mergedElement.getBoundingPoly().getVerticesList().get(3)).build()).build();
                    mergedArray.add(newElement);
                    break;
                }
            }
        }
        return mergedArray;
    }

    private String arrangeWordsInOrder(List<Pair<EntityAnnotation, EntityMetadata>> entityToMetadata, int k) {
        String mergedLine = "";
        List<Match> line = entityToMetadata.get(k).getSecond().getMatch();

        for (Match match : line) {
            int index = match.getMatchLineNum();
            String matchedWordForLine = entityToMetadata.get(index).getFirst().getDescription();

            int mainX = entityToMetadata.get(k).getFirst().getBoundingPoly().getVerticesList().get(0).getX();
            int compareX = entityToMetadata.get(index).getFirst().getBoundingPoly().getVerticesList().get(0).getX();

            if (compareX > mainX) {
                mergedLine = entityToMetadata.get(k).getFirst().getDescription() + ' ' + matchedWordForLine;
            } else {
                mergedLine = matchedWordForLine + ' ' + entityToMetadata.get(k).getFirst().getDescription();
            }
        }
        return mergedLine;
    }

    public int getYMax(EntityAnnotation data) {
        int maxY = Integer.MIN_VALUE;
        for (Vertex vertex : data.getBoundingPoly().getVerticesList()) {
            if (vertex.getY() > maxY) {
                maxY = vertex.getY();
            }
        }
        return maxY;
    }

    private AnnotateImageResponse invertAxis(AnnotateImageResponse data, int yMax) {
        List<EntityAnnotation> newEntities = new ArrayList<>();
        newEntities.add(data.getTextAnnotationsList().get(0));
        for (int i = 1; i < data.getTextAnnotationsList().size(); i++) {
            List<Vertex> vertexList = new ArrayList<>();
            for (Vertex vertex : data.getTextAnnotationsList().get(i).getBoundingPoly().getVerticesList()) {
                vertexList.add(Vertex.newBuilder().mergeFrom(vertex).clearY().setY(yMax - vertex.getY()).build());
            }
            EntityAnnotation.Builder entityBuilder = EntityAnnotation.newBuilder().mergeFrom(data.getTextAnnotationsList().get(i));
            entityBuilder.setBoundingPoly(entityBuilder.getBoundingPolyBuilder().clearVertices().addAllVertices(vertexList).build());
            EntityAnnotation newEntity = entityBuilder.build();
            newEntities.add(newEntity);
        }
        AnnotateImageResponse.Builder responseBuilder = AnnotateImageResponse.newBuilder().mergeFrom(data);
        responseBuilder.clearTextAnnotations();
        responseBuilder.addAllTextAnnotations(newEntities);
        return responseBuilder.build();
    }

    private List<Pair<EntityAnnotation, EntityMetadata>> getBoundingPolygon(List<EntityAnnotation> mergedArray) {
        List<Pair<EntityAnnotation, EntityMetadata>> entityAnnotationToMetadata = new ArrayList<>();
        for (int index = 0; index < mergedArray.size(); index++) {
            EntityAnnotation it = mergedArray.get(index);
            List<Vertex> arr = new ArrayList<>();
            int h1 = it.getBoundingPoly().getVerticesList().get(0).getY() - it.getBoundingPoly().getVerticesList().get(3).getY();
            int h2 = it.getBoundingPoly().getVerticesList().get(1).getY() - it.getBoundingPoly().getVerticesList().get(2).getY();
            int h = h1;
            if (h2 > h1) {
                h = h2;
            }
            int avgHeight = (int) (h * 0.6);

            arr.add(it.getBoundingPoly().getVerticesList().get(1));
            arr.add(it.getBoundingPoly().getVerticesList().get(0));
            Rectangle line1 = getRectangle(arr, avgHeight, true);

            arr = new ArrayList<>();
            arr.add(it.getBoundingPoly().getVerticesList().get(2));
            arr.add(it.getBoundingPoly().getVerticesList().get(3));
            Rectangle line2 = getRectangle(arr, avgHeight, false);

            entityAnnotationToMetadata.add(new Pair<>(it, new EntityMetadata(createPolygon(line1, line2), index, new ArrayList<>(), false)));
        }
        return entityAnnotationToMetadata;
    }

    private void combineBoundingPolygon(List<Pair<EntityAnnotation, EntityMetadata>> entityToMetadata) {
        // select one word from the array
        for (int index1 = 0; index1 < entityToMetadata.size(); index1++) {
            Pair<EntityAnnotation, EntityMetadata> it = entityToMetadata.get(index1);
            Polygon bigBB = it.getSecond().getBigBB();
            // iterate through all the array to find the match
            for (int index2 = index1; index2 < entityToMetadata.size(); index2++) {
                Pair<EntityAnnotation, EntityMetadata> k = entityToMetadata.get(index2);
                // Do not compare with the own bounding box and which was not matched with a line
                if (index1 != index2 && !k.getSecond().isMatched()) {
                    int insideCount = 0;
                    for (Vertex coordinate : k.getFirst().getBoundingPoly().getVerticesList()) {
                        if (bigBB.contains(coordinate.getX(), coordinate.getY())) {
                            insideCount += 1;
                        }
                    }
                    // all four points were inside the big bb
                    if (insideCount == 4) {
                        it.getSecond().getMatch().add(new Match(insideCount, index2));
                        k.getSecond().setMatched(true);
                    }
                }
            }
        }
    }


    private Rectangle getRectangle(List<Vertex> v, double avgHeight, boolean isAdd) {
        double firstCandidate;
        double secondCandidate;
        if (isAdd) {
            secondCandidate = v.get(1).getY() + avgHeight;
            firstCandidate = v.get(0).getY() + avgHeight;
        } else {
            secondCandidate = v.get(1).getY() - avgHeight;
            firstCandidate = v.get(0).getY() - avgHeight;
        }

        double yDiff = secondCandidate - firstCandidate;
        double xDiff = v.get(1).getX() - v.get(0).getX();

        double gradient = yDiff / xDiff;

        int xThreshMin = 1;
        int xThreshMax = 2000;

        double yMin;
        double yMax;
        if (gradient == 0.0) {
            yMin = firstCandidate;
            yMax = firstCandidate;
        } else {
            yMin = (firstCandidate) - (gradient * (v.get(0).getX() - xThreshMin));
            yMax = (firstCandidate) + (gradient * (xThreshMax - v.get(0).getX()));
        }
        yMin = Math.round(yMin);
        yMax = Math.round(yMax);
        return new Rectangle(xThreshMin, xThreshMax, yMin, yMax);
    }

    private Polygon createPolygon(Rectangle line1, Rectangle line2) {
        Polygon polygon = new Polygon();
        polygon.addPoint(line1.getxMin(), (int) Math.round(line1.getyMin()));
        polygon.addPoint(line1.getxMax(), (int) Math.round(line1.getyMax()));
        polygon.addPoint(line2.getxMax(), (int) Math.round(line2.getyMax()));
        polygon.addPoint(line2.getxMin(), (int) Math.round(line2.getyMin()));
        return polygon;
    }

    private Object deepCopy(Object t) {
        String serializedObj = new Gson().toJson(t);
        return new Gson().fromJson(serializedObj, AnnotateImageResponse.class);
    }

    static class Rectangle {
        private int xMin;
        private int xMax;
        private double yMin;
        private double yMax;

        public Rectangle(int xMin, int xMax, double yMin, double yMax) {
            this.xMin = xMin;
            this.xMax = xMax;
            this.yMin = yMin;
            this.yMax = yMax;
        }

        public int getxMin() {
            return xMin;
        }

        public void setxMin(int xMin) {
            this.xMin = xMin;
        }

        public int getxMax() {
            return xMax;
        }

        public void setxMax(int xMax) {
            this.xMax = xMax;
        }

        public double getyMin() {
            return yMin;
        }

        public void setyMin(double yMin) {
            this.yMin = yMin;
        }

        public double getyMax() {
            return yMax;
        }

        public void setyMax(double yMax) {
            this.yMax = yMax;
        }
    }

    static class EntityMetadata {
        private final Polygon bigBB;
        private int lineNum;
        private final List<Match> match;
        private boolean matched;

        public EntityMetadata(Polygon bigBB, int lineNum, List<Match> match, boolean matched) {
            this.bigBB = bigBB;
            this.lineNum = lineNum;
            this.match = match;
            this.matched = matched;
        }

        public Polygon getBigBB() {
            return bigBB;
        }

        public List<Match> getMatch() {
            return match;
        }

        public void setMatched(boolean matched) {
            this.matched = matched;
        }

        public boolean isMatched() {
            return matched;
        }
    }

    static class Match {
        private final int matchCount;
        private final int matchLineNum;

        public Match(int matchCount, int matchLineNum) {
            this.matchCount = matchCount;
            this.matchLineNum = matchLineNum;
        }

        public int getMatchCount() {
            return matchCount;
        }

        public int getMatchLineNum() {
            return matchLineNum;
        }
    }
}

File: src/test/java/org/vaadin/addons/ai/formfiller/utils/GoogleVisionLineSegmentationParser.kt
package org.vaadin.addons.ai.formfiller.utils

import com.google.cloud.vision.v1.AnnotateImageResponse
import com.google.cloud.vision.v1.BoundingPoly
import com.google.cloud.vision.v1.EntityAnnotation
import com.google.cloud.vision.v1.Vertex
import com.google.gson.Gson
import java.awt.Polygon
import kotlin.math.round

// https://github.com/sshniro/line-segmentation-algorithm-to-gcp-vision/blob/master/kotlin/src/main/kotlin/com/google/vision/linesegmentation/GoogleVisionLineSegmentationParser.kt
// https://stackoverflow.com/questions/42391009/text-extraction-line-by-line

class GoogleVisionLineSegmentationParser {
    /**
     * GCP Vision groups several nearby words to appropriate lines
     * But will not group words that are too far away
     * This function combines nearby words and create a combined bounding polygon
     */
    @Suppress("UNCHECKED_CAST")
    fun initLineSegmentation(data: AnnotateImageResponse): List<String> {
        val yMax = getYMax(data.getTextAnnotations(0))

        val newData = invertAxis(data, yMax)
        // The first index refers to the auto identified words which belongs to a sings line
        var lines = newData.textAnnotationsList[0].description.split('\n').toMutableList()
        // gcp vision full text
        var rawText = (deepCopy(newData) as AnnotateImageResponse).textAnnotationsList

        // reverse to use lifo, because array.shift() will consume 0(n)
        lines = lines.reversed().toMutableList()
        rawText = rawText.reversed().toMutableList()
        //to remove the zeroth element which gives the total summary of the text
        rawText.removeAt(rawText.size - 1)

        val mergedArray = getMergedLines(lines, rawText);
        val entityToMetadata = getBoundingPolygon(mergedArray);

        combineBoundingPolygon(entityToMetadata);
        return constructLineWithBoundingPolygon(entityToMetadata);
    }

    // TODO implement the line ordering for multiple words
    protected fun constructLineWithBoundingPolygon(entityToMetadata: List<Pair<EntityAnnotation, EntityMetadata>>): List<String> {
        var finalArray = ArrayList<String>();
        entityToMetadata.forEachIndexed { index, it ->
            if (!it.second.matched) {
                if (it.second.match.size == 0) {
                    finalArray.add(it.first.description)
                } else {
                    // arrangeWordsInOrder(mergedArray, i);
                    // let index = mergedArray[i]['match'][0]['matchLineNum'];
                    // let secondPart = mergedArray[index].description;
                    // finalArray.push(mergedArray[i].description + ' ' +secondPart);
                    finalArray.add(arrangeWordsInOrder(entityToMetadata, index));
                }
            }
        }
        return finalArray;
    }

    private fun getMergedLines(lines: MutableList<String>, rawText: MutableList<EntityAnnotation>): List<EntityAnnotation> {
        val mergedArray = ArrayList<EntityAnnotation>();
        while (lines.size != 1) {
            var l = lines.removeAt(lines.size - 1);
            val l1 = l
            var status = true;
            var mergedElement: EntityAnnotation? = null

            while (true) {
                if (rawText.isEmpty()) {
                    break;
                }
                val wElement = rawText.removeAt(rawText.size - 1);
                val w = wElement.description;

                val index = l.indexOf(w);

                l = l.substring(index + w.length);

                if (status) {
                    status = false;
                    // set starting coordinates
                    mergedElement = wElement;
                }
                if (l == "") {
                    val newElement = EntityAnnotation.newBuilder().mergeFrom(mergedElement)
                            .setDescription(l1)
                            .setBoundingPoly(BoundingPoly.newBuilder().mergeFrom(mergedElement!!.boundingPoly).clearVertices()
                                    .addVertices(0, mergedElement.boundingPoly.verticesList[0])
                                    .addVertices(1, wElement.boundingPoly.verticesList[1])
                                    .addVertices(2, wElement.boundingPoly.verticesList[2])
                                    .addVertices(3, mergedElement.boundingPoly.verticesList[3]).build()).build()
                    mergedArray.add(newElement);
                    break;
                }
            }
        }
        return mergedArray;
    }

    private fun arrangeWordsInOrder(entityToMetadata: List<Pair<EntityAnnotation, EntityMetadata>>, k: Int): String {
        var mergedLine = "";
        var line = entityToMetadata[k].second.match;

        line.forEach {
            val index = it.matchLineNum;
            val matchedWordForLine = entityToMetadata[index].first.description;

            val mainX = entityToMetadata[k].first.boundingPoly.verticesList[0].x;
            val compareX = entityToMetadata[index].first.boundingPoly.verticesList[0].x;

            if (compareX > mainX) {
                mergedLine = entityToMetadata[k].first.description + ' ' + matchedWordForLine;
            } else {
                mergedLine = matchedWordForLine + ' ' + entityToMetadata[k].first.description;
            }
        }
        return mergedLine;
    }

    /**
     * @Method computes the maximum y coordinate from the identified text blob
     * @param data
     * @returns {*}
     */
    fun getYMax(data: EntityAnnotation): Int {
        var maxY = Int.MIN_VALUE
        for (vertex in data.boundingPoly.verticesList) {
            if (vertex.y > maxY) {
                maxY = vertex.y
            }
        }
        return maxY
    }

    /**
     * @Method inverts the y axis coordinates for easier computation
     * as the google vision starts the y axis from the bottom
     * @param data
     * @param yMax
     * @returns {*}
     */
    private fun invertAxis(data: AnnotateImageResponse, yMax: Int): AnnotateImageResponse {
        //TODO Don't think this is needed
        //data = fillMissingValues(data);
        val newEntities = ArrayList<EntityAnnotation>()
        newEntities.add(data.textAnnotationsList[0])
        for (i in 1 until data.textAnnotationsList.size) {
            val vertexList = ArrayList<Vertex>()
            data.textAnnotationsList[i].boundingPoly.verticesList.forEach {
                vertexList.add(Vertex.newBuilder().mergeFrom(it).clearY().setY(yMax - it.y).build())
            }
            val entityBuilder = EntityAnnotation.newBuilder().mergeFrom(data.textAnnotationsList[i])
            entityBuilder.boundingPoly = entityBuilder.boundingPolyBuilder.clearVertices().addAllVertices(vertexList).build()
            val newEntity = entityBuilder.build()
            newEntities.add(newEntity)
        }
        val responseBuilder = AnnotateImageResponse.newBuilder().mergeFrom(data)
        responseBuilder.clearTextAnnotations()
        responseBuilder.addAllTextAnnotations(newEntities)
        return responseBuilder.build()
    }


    /**
     *
     * @param mergedArray
     */
    private fun getBoundingPolygon(mergedArray: List<EntityAnnotation>): List<Pair<EntityAnnotation, EntityMetadata>> {
        val entityAnnotationToMetadata = ArrayList<Pair<EntityAnnotation, EntityMetadata>>()
        mergedArray.forEachIndexed { index, it ->
            var arr = ArrayList<Vertex>();
            // calculate line height
            val h1 = it.boundingPoly.verticesList[0].y - it.boundingPoly.verticesList[3].y;
            val h2 = it.boundingPoly.verticesList[1].y - it.boundingPoly.verticesList[2].y;
            var h = h1;
            if (h2 > h1) {
                h = h2
            }
            val avgHeight = h * 0.6;

            arr.add(it.boundingPoly.verticesList[1]);
            arr.add(it.boundingPoly.verticesList[0]);
            val line1 = getRectangle(arr, avgHeight, true);

            arr = ArrayList();
            arr.add(it.boundingPoly.verticesList[2]);
            arr.add(it.boundingPoly.verticesList[3]);
            val line2 = getRectangle(arr, avgHeight, false);

            entityAnnotationToMetadata.add(Pair(it, EntityMetadata(createPolygon(line1, line2), index, ArrayList(), false)))
        }
        return entityAnnotationToMetadata
    }

    private fun combineBoundingPolygon(entityToMetadata: List<Pair<EntityAnnotation, EntityMetadata>>) {
        // select one word from the array
        entityToMetadata.forEachIndexed { index1, it ->
            val bigBB = it.second.bigBB;
            // iterate through all the array to find the match
            for (index2 in index1 until entityToMetadata.size) {
                val k = entityToMetadata[index2]
                // Do not compare with the own bounding box and which was not matched with a line
                if (index1 != index2 && !k.second.matched) {
                    var insideCount = 0;
                    k.first.boundingPoly.verticesList.forEach { coordinate ->
                        if (bigBB.contains(coordinate.x, coordinate.y)) {
                            insideCount += 1;
                        }
                    }
                    // all four point were inside the big bb
                    if (insideCount == 4) {
                        it.second.match.add(Match(insideCount, index2))
                        k.second.matched = true;
                    }

                }
            }
        }
    }

    private fun getRectangle(v: MutableList<Vertex>, avgHeight: Double, isAdd: Boolean): Rectangle {
        val firstCandidate: Double
        val secondCandidate: Double
        if (isAdd) {
            secondCandidate = v[1].y + avgHeight
            firstCandidate = v[0].y + avgHeight
        } else {
            secondCandidate = v[1].y - avgHeight
            firstCandidate = v[0].y - avgHeight
        }

        val yDiff = (secondCandidate - firstCandidate);
        val xDiff = (v[1].x - v[0].x);

        val gradient = yDiff / xDiff;

        val xThreshMin = 1;
        val xThreshMax = 2000;

        var yMin: Double
        var yMax: Double
        if (gradient == 0.0) {
            // extend the line
            yMin = firstCandidate;
            yMax = firstCandidate;
        } else {
            yMin = (firstCandidate) - (gradient * (v[0].x - xThreshMin));
            yMax = (firstCandidate) + (gradient * (xThreshMax - v[0].x));
        }
        yMin = round(yMin);
        yMax = round(yMax);
        return Rectangle(xThreshMin, xThreshMax, yMin, yMax)
    }

    private fun createPolygon(line1: Rectangle, line2: Rectangle): Polygon {
        val polygon = Polygon()
        polygon.addPoint(line1.xMin, round(line1.yMin).toInt())
        polygon.addPoint(line1.xMax, round(line1.yMax).toInt())
        polygon.addPoint(line2.xMax, round(line2.yMax).toInt())
        polygon.addPoint(line2.xMin, round(line2.yMin).toInt())
        return polygon
    }

    private fun deepCopy(t: Any): Any {
        val serializedObj = Gson().toJson(t);
        return Gson().fromJson(serializedObj, AnnotateImageResponse::class.java)
    }

    data class Rectangle(
            val xMin: Int,
            val xMax: Int,
            val yMin: Double,
            val yMax: Double
    )

    data class EntityMetadata(
            val bigBB: Polygon,
            val lineNum: Int,
            val match: MutableList<Match>,
            var matched: Boolean
    )

    data class Match(
            val matchCount: Int,
            val matchLineNum: Int
    )
}